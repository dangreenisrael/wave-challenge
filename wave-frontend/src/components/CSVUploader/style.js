export default {
  margin: 10
};
