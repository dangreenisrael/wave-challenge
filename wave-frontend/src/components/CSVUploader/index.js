import React, {Component} from 'react';
import fetcher from '../../modules/fetcher';
import style from './style';
import PropTypes from 'prop-types';

class CSVUpload extends Component {
  constructor(props) {
    super(props);
    this.submit = this.submit.bind(this);
    this.state = {inputType: 'file'};
  }
  submit({target}) {
    const {reloadData} = this.props;
    const [file] = target.files;
    const data = new FormData();
    data.append('timesheet', file);
    fetcher
      .post('timesheets', data)
      .then(() => {
        this.setState({inputType: ''}); //Reset file input
        reloadData();
        this.setState({inputType: 'file'}); //Reset file input
      })
      .catch(error => {
        const {status} = error.response;
        this.setState({inputType: ''}); //Reset file input
        if (status === 409) {
          alert('You submitted a report that we have already processed');
        } else if (status === 400) {
          alert("You submitted a file that isn't a CSV");
        } else {
          alert('Whoops, There seems to be a problem with our server');
        }
        this.setState({inputType: 'file'}); //Reset file input
      });
  }
  render() {
    const {inputType} = this.state;
    return (
      <label className="btn btn-primary btn-lg" style={style}>
        Upload a Time Report (CSV)
        <input type={inputType} style={{display: 'none'}} onChange={this.submit} accept=".csv" />
      </label>
    );
  }
}
CSVUpload.propTypes = {
  reloadData: PropTypes.func.isRequired
};

export default CSVUpload;
