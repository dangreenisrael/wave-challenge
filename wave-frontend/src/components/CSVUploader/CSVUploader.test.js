import React from 'react';
import renderer from 'react-test-renderer';
import CSVUploader from '.';
import {shallow} from 'enzyme';

test('CSVUploader Renders Correctly', () => {
  const reloadData = jest.fn();
  const tree = renderer.create(<CSVUploader reloadData={reloadData} />).toJSON();
  expect(tree).toMatchSnapshot();
});

/*
 Note: The calls to the mock will go in this order:
 200
 400
 409
 500
 */

test('Handles a successful file upload', done => {
  global.alert = jest.fn();
  const reloadData = jest.fn();
  const wrapper = shallow(<CSVUploader reloadData={reloadData} />);
  const f = new File(['hello'], 'filename');
  wrapper.instance().submit({
    target: {files: [f]}
  });

  wrapper.instance().submit({
    target: {files: [f]}
  });

  wrapper.instance().submit({
    target: {files: [f]}
  });

  wrapper.instance().submit({
    target: {files: [f]}
  });

  setTimeout(
    () => {
      const [[e400], [e409], [e500]] = global.alert.mock.calls;
      expect(global.alert).toHaveBeenCalledTimes(3);
      expect(e400).toBe("You submitted a file that isn't a CSV");
      expect(e409).toBe('You submitted a report that we have already processed');
      expect(e500).toBe('Whoops, There seems to be a problem with our server');
      done();
    },
    1000
  );
});
