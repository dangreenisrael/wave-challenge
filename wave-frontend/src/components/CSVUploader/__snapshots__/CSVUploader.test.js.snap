// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSVUploader Renders Correctly 1`] = `
<label
  className="btn btn-primary btn-lg"
  style={
    Object {
      "margin": 10,
    }
  }
>
  Choose a CSV
  <input
    accept=".csv"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
</label>
`;
